src/java.base/share/classes/java/lang/classfile/TypeKind.java

```diff
@@ -115,35 +123,51 @@ public static TypeKind fromNewArrayCode(int newarraycode) {
             case 9 -> TypeKind.ShortType;
             case 10 -> TypeKind.IntType;
             case 11 -> TypeKind.LongType;
-            default -> throw new IllegalArgumentException("Bad new array code: " + newarraycode);
+            default -> throw new IllegalArgumentException("Bad newarray code: " + newarrayCode);
         };
     }
 
     /**
      * {@return the type kind associated with the specified field descriptor}
      * @param s the field descriptor
+     * @throws IllegalArgumentException only if the descriptor is not valid
      */
     public static TypeKind fromDescriptor(CharSequence s) {
-        return switch (s.charAt(0)) {
-            case '[', 'L' -> TypeKind.ReferenceType;
-            case 'B' -> TypeKind.ByteType;
-            case 'C' -> TypeKind.CharType;
-            case 'Z' -> TypeKind.BooleanType;
-            case 'S' -> TypeKind.ShortType;
-            case 'I' -> TypeKind.IntType;
-            case 'F' -> TypeKind.FloatType;
-            case 'J' -> TypeKind.LongType;
-            case 'D' -> TypeKind.DoubleType;
-            case 'V' -> TypeKind.VoidType;
-            default -> throw new IllegalArgumentException("Bad type: " + s);
-        };
```

Comment:
This maps to a `tableswitch`, could you explain why a statically initialized array addressed with a custom hash code suppose to be faster?
Thanks.

Reply:
I actually am not sure of the reason; it is indeed a tableswitch in bytecode. @cl4es might know it better?

Reply:
Generally I agree with the API changes and with the `descriptor.isPrimitive()` test before requesting a descriptor string.
However I'm missing the point of the other changes, which only replace this single `tableswitch` with something way more complicated, without clear performance gain.

Reply:
The code I tested saw a decidedly larger speed-up, perhaps hinging on other optimizations such as being able to avoid the `s.isEmpty()` check (could maybe catch IIOBE) and not matching on `L` and `[` (the lookup table only needs to deal with primitives). As it looks here I agree the speed-up is negligible and not worth it, and since `TypeDescriptor.OfField` isn't a sealed hierarchy perhaps we need some redundant checks.